File: _Enutrof_[__-1]🎧🌝💉☕/Day 39/find-right-interval.py
class Solution:
    def binarySearch(self, end):
        l, r = 0, len(self.store)-1
        res = -1
        while l <= r:
            mid = l+(r-l)//2
            if self.store[mid][0][0]>=end:
                res = self.store[mid][1]
                r = mid-1
            else:
                l = mid+1
        return res
    def findRightInterval(self, intervals: List[List[int]]) -> List[int]:
        self.store =sorted([[intervals,i] for i,intervals in enumerate(intervals)])
        res=[]
        for start, end in intervals:
            res.append(self.binarySearch(end))
        return res
      

File: olix/day-39/find-right-interval.js
/**
 * @param {number[][]} intervals
 * @return {number[]}
 */
var findRightInterval = function (intervals) {
  let result = [];
  let sorted = intervals.sort((a, b) => a[0] - b[0]);
  for (let i = 0; i < intervals.length; i++) {
    let start = intervals[i][0];
    let end = intervals[i][1];
    let index = binarySearch(sorted, end);
    result.push(index);
  }
  return result;
};

function binarySearch(arr, target) {
  let start = 0;
  let end = arr.length - 1;
  while (start <= end) {
    let mid = Math.floor((start + end) / 2);
    console.log(mid);
    if (arr[mid][0] === target) {
      return mid;
    } else if (arr[mid][0] > target) {
      end = mid - 1;
    } else {
      start = mid + 1;
    }
  }
  return -1;
}

console.log(
  binarySearch([
    [1, 4],
    [2, 3],
    [3, 4],
  ])
);
